### PR TITLE
bootstrap ecash module

### PR DESCRIPTION
**Pull request by plebhash**
Adds the `ecash` module and initializes a mint service on `main.rs`.
The mint service is initialized with info coming from a Config file (`config.toml` as default).

Mostly inspired by @thesimplekid's WIP on [`cashu-rs-mint`](https://github.com/thesimplekid/cashu-rs-mint), which is a mint based on [`cdk`](https://github.com/cashubtc/cdk)

For extra context, here's some ongoing brainstorming: https://github.com/plebemineira/plebpool/discussions/7

### CODE REVIEW

File: src/ecash/config.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+fn max_order_default() -> u8 {
+    32
+}
+
+fn last_pay_path() -> String {
+    "/tmp/plebpool/last_pay".to_string()
+}
+
```
Comment (storopoli) [inline src/ecash/config.rs:55]:
```suggestion
// assumes /tmp/plebpool exists
// todo: unify all paths into one single root path
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct MintConfig {
    pub url: String,
    pub db_path: PathBuf,
    pub last_pay_path: String,
    pub listen_host: String,
    pub listen_port: u16,
    pub mnemonic: String,
    pub derivation_path: String,
    pub max_order: u8,
    pub min_fee_reserve: cdk::Amount,
    pub min_fee_percent: f32,
    pub ln: LnConfig,
}

impl Default for MintConfig {
    fn default() -> Self {
        db_path: PathBuf::from_str("/tmp/plebpool/mint.redb").unwrap(),
        derivation_path: "0/0/0/0".to_string(),
        max_order: 32,
        last_pay_path: "/tmp/plebpool/last_pay".to_string(),
        .. Default::default()
    }
}
```

Ref: https://doc.rust-lang.org/std/default/trait.Default.html

File: src/ecash/service/mod.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    .await?;
+
+    debug!("Mint created");
+
+    let cln_socket = expand_path(
+        ecash_config
+            .ln
+            .cln_path
+            .clone()
+            .ok_or(anyhow::anyhow!("cln socket not defined"))?
```
Comment (douglaz) [inline src/ecash/service/mod.rs:61]: `use anyhow::Context;`

```suggestion
            .context("cln socket not defined")?
```
and so on...

File: src/ecash/service/mod.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    mint: Arc<tokio::sync::Mutex<cdk::mint::Mint>>,
+}
+
+pub async fn mint_service(
+    config_file_arg: String,
+) -> anyhow::Result<tokio::task::JoinHandle<()>> {
+    let ecash_config = match ecash::config::EcashConfig::new(&Some(config_file_arg)) {
+        Ok(mint_settings) => mint_settings,
+        Err(e) => {
+            error!("{}. Halting.", e);
```
Comment (douglaz) [inline src/ecash/service/mod.rs:30]:
```suggestion
            error!("{e:?}. Halting.");
```

Comment (douglaz) [reply src/ecash/service/mod.rs:30]: `{:?} ` on errors will bring the stacktrace

File: src/ecash/service/mod.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    .ok_or(anyhow::anyhow!("cln socket not defined"))?;
+
+    let last_pay_path = PathBuf::from_str(&ecash_config.mint.last_pay_path.clone())?;
+
+    // Create last_pay_path if it doesn't exist
+    match fs::metadata(&last_pay_path) {
+        Ok(_) => (),
+        Err(_e) => {
+            // Create the parent directory if it doesn't exist
+            fs::create_dir_all(last_pay_path.parent().unwrap())?;
```
Comment (douglaz) [inline src/ecash/service/mod.rs:74]:
```suggestion
            fs::create_dir_all(last_pay_path.parent()?)?;
```

Comment (plebhash) [reply src/ecash/service/mod.rs]: simply replacing with `?` breaks compilation.

c5e916fad5929702fa1b7827c123d63169bef5a7 did the trick

File: src/ecash/service/mod.rs (hunk 2 above)
Comment (plebhash) [reply src/ecash/service/mod.rs]: 98b7ad102b4551eee2fa013caecb3c07013b6dc5

File: src/ecash/service/mod.rs (hunk 3 above)
Comment (plebhash) [reply src/ecash/service/mod.rs]: 664863488df1ca8702a98f42e38fb9edeecfaa8d

Comment (douglaz) [reply src/ecash/service/mod.rs]: I'm sorry, you also need to use .context, something like:
```rust
let parent_path = last_pay_path
                .parent()
                .context("Parent directory not found")?;
fs::create_dir_all(parent_path)?;
```

Comment (douglaz) [reply src/ecash/service/mod.rs]: Or to give a better context to the caller:
```rust
let parent_path = last_pay_path
                .parent()
                .with_context(|| format!("Parent directory not found for {last_pay_path:?}"))?;
fs::create_dir_all(parent_path)?;
```

File: src/ecash/service/mod.rs (hunk 4 above)
Comment (douglaz) [reply src/ecash/service/mod.rs]: I'm sorry, you also need to use .context, something like:
```rust
let parent_path = last_pay_path
                .parent()
                .context("Parent directory not found")?;
fs::create_dir_all(parent_path)?;
```

Or to give a better context to the caller:
```rust
let parent_path = last_pay_path
                .parent()
                .with_context(|| format!("Parent directory not found for {last_pay_path:?}"))?;
fs::create_dir_all(parent_path)?;
```

**Pull request merged**